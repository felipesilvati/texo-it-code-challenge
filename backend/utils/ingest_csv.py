import pandas as pd
from models import Movie, Studio, Producer, Award

def ingest_csv_to_db(csv_path, app, db):
    with app.app_context():
        # Read the CSV file into a DataFrame
        df = pd.read_csv(csv_path, delimiter=';', encoding='utf-8')

        # Iterate over the DataFrame
        for _, row in df.iterrows():
            movie = Movie(title=row['title'], year=row['year'])
            db.session.add(movie)

            # Handle studios
            if not pd.isna(row['studios']):
                studios = row['studios'].split(',')
                for studio_name in studios:
                    studio_name = studio_name.strip()
                    studio = Studio.query.filter_by(name=studio_name).first()
                    if not studio:
                        studio = Studio(name=studio_name)
                        db.session.add(studio)
                    movie.studios.append(studio)

            # Handle producers
            if not pd.isna(row['producers']):
                producers = row['producers'].split(',')
                for producer_name in producers:
                    producer_name = producer_name.strip()
                    producer = Producer.query.filter_by(name=producer_name).first()
                    if not producer:
                        producer = Producer(name=producer_name)
                        db.session.add(producer)
                    movie.producers.append(producer)

            # Handle winner
            is_winner = False
            if 'winner' in row and not pd.isna(row['winner']) and row['winner'].lower().strip() == 'yes':
                is_winner = True
            award = Award(is_winner=is_winner)
            movie.awards.append(award)

            db.session.commit()