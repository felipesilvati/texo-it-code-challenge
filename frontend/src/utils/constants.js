export const BASE_API_URL = 'https://tools.texoit.com/backend-java/api/movies';