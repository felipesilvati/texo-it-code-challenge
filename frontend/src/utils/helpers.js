import { message } from 'antd';
export const constructQueryString = (params) => {
  return '?' + Object.keys(params)
    .filter(key => params[key] !== null && params[key] !== undefined)
    .map(key => `${encodeURIComponent(key)}=${encodeURIComponent(params[key])}`)
    .join('&');
};

export const onError = (error) => {
  message.error('Failed to load data')
  console.error(error);
};