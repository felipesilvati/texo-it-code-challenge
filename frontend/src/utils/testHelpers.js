import { QueryClient } from '@tanstack/react-query';

export const createTestQueryClient = () => new QueryClient({
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: process.env.NODE_ENV === 'test' ? () => { } : console.error,
  },
  defaultOptions: {
    queries: {
      retry: false,
    },
  },
});