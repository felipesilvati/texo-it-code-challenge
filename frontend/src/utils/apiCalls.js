import axios from 'axios';
import { BASE_API_URL } from '@/utils/constants';
import { constructQueryString } from '@/utils/helpers';

export const fetchMovies = ({ queryKey }) => {
  const [_, page, size, filters] = queryKey;
  const queryString = constructQueryString({ page, size, ...filters });
  const url = `${BASE_API_URL}${queryString}`;
  return axios.get(url).then(res => res.data);
};

export const fetchYearsWithMultipleWinners = () => {
  return axios
    .get(`${BASE_API_URL}/?projection=years-with-multiple-winners`)
    .then((res) => res.data?.years)
};

export const fetchTopThreeStudiosWithWinners = () => {
  return axios
    .get(`${BASE_API_URL}/?projection=studios-with-win-count`)
    .then(res => res.data?.studios?.slice(0, 3))
};