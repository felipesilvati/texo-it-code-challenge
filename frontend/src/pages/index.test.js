import { render } from '@testing-library/react';
import { QueryClientProvider } from '@tanstack/react-query';
import { useRouter } from 'next/router';
import Dashboard from '@/pages/index';
import { createTestQueryClient } from '@/utils/testHelpers';

const YearsWithMultipleWinnersMock = () => <div data-testid="YearsWithMultipleWinners" />;
YearsWithMultipleWinnersMock.displayName = 'YearsWithMultipleWinnersMock';
jest.mock('@/components/YearsWithMultipleWinners', () => YearsWithMultipleWinnersMock);

const TopThreeStudiosWithWinnersMock = () => <div data-testid="TopThreeStudiosWithWinners" />;
TopThreeStudiosWithWinnersMock.displayName = 'TopThreeStudiosWithWinnersMock';
jest.mock('@/components/TopThreeStudiosWithWinners', () => TopThreeStudiosWithWinnersMock);

const ProducersWithLongestAndShortestMock = () => <div data-testid="ProducersWithLongestAndShortest" />;
ProducersWithLongestAndShortestMock.displayName = 'ProducersWithLongestAndShortestMock';
jest.mock('@/components/ProducersWithLongestAndShortest', () => ProducersWithLongestAndShortestMock);

const ListMovieWinnersByYearMock = () => <div data-testid="ListMovieWinnersByYear" />;
ListMovieWinnersByYearMock.displayName = 'ListMovieWinnersByYearMock';
jest.mock('@/components/ListMovieWinnersByYear', () => ListMovieWinnersByYearMock);

const LayoutMock = ({ children }) => <div>{children}</div>;
LayoutMock.displayName = 'LayoutMock';
jest.mock('@/components/Layout', () => LayoutMock);


jest.mock('next/router', () => ({
  useRouter: jest.fn(),
}));

const mockPush = jest.fn();
useRouter.mockImplementation(() => ({
  push: mockPush,
  pathname: '/',
}));

function renderDashboard() {
  const queryClient = createTestQueryClient();
  return render(
    <QueryClientProvider client={queryClient}>
      <Dashboard />
    </QueryClientProvider>
  );
}

describe('Dashboard', () => {
  it('renders child components correctly', () => {
    const { getByTestId } = renderDashboard();

    expect(getByTestId('YearsWithMultipleWinners')).toBeInTheDocument();
    expect(getByTestId('TopThreeStudiosWithWinners')).toBeInTheDocument();
    expect(getByTestId('ProducersWithLongestAndShortest')).toBeInTheDocument();
    expect(getByTestId('ListMovieWinnersByYear')).toBeInTheDocument();
  });
});
