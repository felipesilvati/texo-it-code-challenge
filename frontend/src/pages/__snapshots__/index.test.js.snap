// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="ant-flex css-dev-only-do-not-override-1uweeqc ant-flex-gap-large"
    >
      <div
        data-testid="YearsWithMultipleWinners"
      />
      <div
        data-testid="TopThreeStudiosWithWinners"
      />
    </div>
    <div
      class="ant-flex css-dev-only-do-not-override-1uweeqc ant-flex-gap-large"
    >
      <div
        data-testid="ProducersWithLongestAndShortest"
      />
      <div
        data-testid="ListMovieWinnersByYear"
      />
    </div>
  </div>
</DocumentFragment>
`;
